### PR TITLE
[Module][WIP] registration tractogram

### PR DESCRIPTION
**Pull request by anroy1**
## Checklist before requesting a review

- Create the tool:
  - [x] Edit `./modules/nf-scil/<category>/<tool>/main.nf`
  - [x] Edit `./modules/nf-scil/<category>/<tool>/meta.yml`
- Generate the tests:
  - [x] Edit `./tests/modules/nf-scil/<category>/<tool>/main.nf`
  - [x] Edit `./tests/modules/nf-scil/<category>/<tool>/nextflow.config`
  - [x] Add test data locally for tests with the fork repository
  - [x] Generate the test infrastructure and _md5sum_ for all outputs
- Ensure the syntax is correct :
  - [x] Check indentation abides with the rest of the library (don't hesitate to correct others !)
  - [x] Lint everything. Ensure your variables have good names.

### CODE REVIEW

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:22]: I'm not sure you need $ for the first ref : 
def reference = ref ? "--reference $ref" : ""

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:20]: Remove it

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:24]: Be careful here these are store_true options:
1. you probably don't need to add this to the line: "--inverse " + task.ext.inverse
     def inverse = task.ext.inverse ? "--reverse " : ""
2. I think you need to put ifs for these defs (to be confirmed with Alex)

see: https://github.com/scilus/nf-scil/blob/main/modules/nf-scil/reconst/fodf/main.nf

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:32]: The cut_invalid, remove_invalid and keep_invalid are mutually exclusive, not sure if you can manage them like that.

can be an option like:
def invalid = task.ext.invalid_action? task.ext.invalid_action: "--remove_invalid" to have default behavior?

to validate with Alex

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:40]: I think there would be an advantage in removing the loop from the module, otherwise you lose the advantage of nextflow, but it may generate too many files. To see with Alex and Arnaud.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:70]: remove it

File: modules/nf-scil/registration/tractogram/main.nf
Comment (Manonedde) [inline modules/nf-scil/registration/tractogram/main.nf:73]: As written your module will output all the centroids, not just one. Which goes back to my previous comment.

**Review by Manonedde — CHANGES_REQUESTED** [review]
It's a good job but I think we need to rethink a little how the module works and how to manage store_true and mutually_exclusive.

Should discuss it with @AlexVCaron  and @arnaudbore.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [reply modules/nf-scil/registration/tractogram/main.nf:24]: It is not a requirement to use `if/else` statements, the ternary operator does the job (we'll uniformize the modules in the Code freeze period that's coming). As for the boolean flags (what you called store_true), you are correct @Manonedde.

On that matter @anroy1, I want you to be extra careful with those. It's a mistake I've been seeing on a lot of your submitted PRs.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [reply modules/nf-scil/registration/tractogram/main.nf:40]: Yes' for now, we're keeping the loop, as you said for the number of files argument. When we integrate multiprocessing in the modules, we'll parallelize this loop (using parallel or some other bash utility).

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [reply modules/nf-scil/registration/tractogram/main.nf:73]: You indeed have access to each centroid present in the directory here. Use a for loop to cycle on them and `touch` the files you need to create.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [inline modules/nf-scil/registration/tractogram/main.nf:25]: A deformation is a Warp field (.nii.gz) from ANTs. It should be in the `input` section, not `task.ext`

File: tests/modules/nf-scil/registration/tractogram/nextflow.config
Comment (AlexVCaron) [inline tests/modules/nf-scil/registration/tractogram/nextflow.config:2]: remove this line, else you include files from LOAD_TEST_DATA in your test outputs

File: tests/modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [inline tests/modules/nf-scil/registration/tractogram/main.nf:7]: Why the 1 at the end of your module's name ? You don't need to rename the module here, just import it.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [inline modules/nf-scil/registration/tractogram/main.nf:1]: This module needs to change name. What it does is bundle registration, am I wrong ?

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Some comments to answer. It's coming on well !

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [reply modules/nf-scil/registration/tractogram/main.nf:1]: I'll update the sheet, let's name this REGISTRATION_BUNDLEREGISTRATION.

Since to register tractograms, users should use #74 with the correct transforms, not this PR. Thanks a lot !

**Comment by anroy1** [comment]
@AlexVCaron ready for review (i believe)

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Code wise, I don't think I have anything to say, very good job ! I'd like to discuss some points more on the usability side.

First, I think we should replace the `centroid_dir` input and consider taking a list of files instead. There is a way to have them placed in a directory in the process, using the `stageAs` parameter of a path :

```
inputs:
    path(centroid_dir, stageAs: 'centroids/*')
```

You'll have to adapt some of the code, most of it will stay the same though. But the end-user usability would greatly profit, it is quite hard to get files in the same folder in groovy and adding a process just for that is kind of a waste.

Second, if I'm not mistaken, this module is able to transform anything in a trk/tck format, even vtk if it is save correctly. I was wondering if it would be better to replace the `centroid` nomenclature used in this module for `bundle`. Or maybe even drop centroid and bundle for a more general name. Tracts ? Streamlines ? What is your idea on that ?

@arnaudbore @frheault, I'd also like your input on this point. Better make the naming clear now.

**Comment by frheault** [comment]
I believe supporting TRK and TCK would be important (VTK can be dealt with later if necessary)

Also I think the best nomenclature would be 'tractogram' (one or one million streamlines, whole brain or single bundle, that's a tractogram)

**Comment by AlexVCaron** [comment]
Alright. After some considerations, we'll switch back the name to `REGISTRATION_TRACTOGRAM`. Very sorry @anroy1 :cry: I made you work for nothing.

Whilst it's true we could specialize it further for `bundles`, I don't feel the need to. And it does provide the cleanup step in comparison to using `REGISTRATION_ANTSAPPLYTRANSFORMS` (which by the way @frheault does work out-of-the-box on tractograms, I made sure of that !).

**Comment by anroy1** [comment]
@AlexVCaron No problem at all! Do you expect this name change with a modification to the code's logic? Actually, it is really design with the "transform centroids" logic from rbx flow with:

` for centroid in ${centroids_dir};
        do bname=\${centroid/_centroid/}
        bname=\$(basename \${bname} .trk)`

Do you think of an approach to generalize the tractogram input management, from a single tractogram to the special case of rbx' atlas centroids?

**Comment by anroy1** [comment]
Ready for review

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [inline modules/nf-scil/registration/tractogram/main.nf:38]: This way of extracting the output name from the tractogram is a bit outdated. I'm not sure it'll continue to apply consistently. A better way would be to use the tractogram name in full as output name. Since all input tractograms are placed in a directory, there is no overwritting problem : `bname=\${tractogram}`

However, to remain consistent, the output names do have to start by the `prefix` defined in the script. A tractogram could start with `prefix`, so you have to remove it if it is there : `bname=\${tractogram/${prefix}_/_}` (if it's not, the command is still valid, no need to put a condition around to validate).

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
One minor comment. If I'm not clear, let's call each other ! Then, this should be all good to go :ship:

File: modules/nf-scil/registration/tractogram/main.nf
Comment (AlexVCaron) [inline modules/nf-scil/registration/tractogram/main.nf:71]:
```suggestion
        do bname=\${tractogram/${prefix}_/_}
```

**Review by AlexVCaron — CHANGES_REQUESTED** [review]

**Comment by anroy1** [comment]
done! Sorry for delay!

File: modules/nf-scil/registration/tractogram/main.nf
Comment (arnaudbore) [inline modules/nf-scil/registration/tractogram/main.nf:57]: no need for reference here

**Review by arnaudbore — COMMENTED** [review]
A couple of comments. Will merge right after @anroy1 answered them. Then a new PR with updated tests will come up quickly.

File: modules/nf-scil/registration/tractogram/main.nf
Comment (anroy1) [reply modules/nf-scil/registration/tractogram/main.nf:42]: The \$ext approch was suggested by francois as managing trk and tck. But I assume you want all outputs to be only trk?

Comment (arnaudbore) [reply modules/nf-scil/registration/tractogram/main.nf:42]: this a temporary trk does not need to be tck here.

**Pull request merged**
